// box-node-sdk/lib/managers/shield-information-barrier-reports.generated.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var url_path_1 = __importDefault(require("../util/url-path"));
/**
 */
var ShieldInformationBarrierReportsManager = /** @class */ (function () {
    /**
     * @param {BoxClient} client The Box API Client that is responsible for making calls to the API
     */
    function ShieldInformationBarrierReportsManager(client) {
        this.client = client;
    }
    /**
     * Get shield information barrier report by ID
     *
     * Retrieves a shield information barrier report by its ID.
     * @param {object} options Options for the request
     * @param {string} options.shield_information_barrier_report_id The ID of the shield information barrier Report.
     * @param {Function} [callback] Passed the result if successful, error otherwise
     * @returns {Promise<schemas.ShieldInformationBarrierReport>} A promise resolving to the result or rejecting with an error
     */
    ShieldInformationBarrierReportsManager.prototype.getById = function (options, callback) {
        var shieldInformationBarrierReportId = options.shield_information_barrier_report_id, queryParams = __rest(options, ["shield_information_barrier_report_id"]), apiPath = (0, url_path_1.default)('shield_information_barrier_reports', shieldInformationBarrierReportId), params = {
            qs: queryParams,
        };
        return this.client.wrapWithDefaultHandler(this.client.get)(apiPath, params, callback);
    };
    /**
     * List shield information barrier reports
     *
     * Lists shield information barrier reports with specific IDs.
     * @param {object} options Options for the request
     * @param {string} options.shield_information_barrier_id The ID of the shield information barrier.
     * @param {string} [options.marker] Defines the position marker at which to begin returning results. This is used when paginating using marker-based pagination. This requires `usemarker` to be set to `true`.
     * @param {number} [options.limit] The maximum number of items to return per page.
     * @param {Function} [callback] Passed the result if successful, error otherwise
     * @returns {Promise<object>} A promise resolving to the result or rejecting with an error
     */
    ShieldInformationBarrierReportsManager.prototype.getAll = function (options, callback) {
        var queryParams = __rest(options, []), apiPath = (0, url_path_1.default)('shield_information_barrier_reports'), params = {
            qs: queryParams,
        };
        return this.client.wrapWithDefaultHandler(this.client.get)(apiPath, params, callback);
    };
    /**
     * Create shield information barrier report
     *
     * Creates a shield information barrier report for a given barrier.
     * @param {schemas.ShieldInformationBarrierReference} body
     * @param {object} [options] Options for the request
     * @param {Function} [callback] Passed the result if successful, error otherwise
     * @returns {Promise<schemas.ShieldInformationBarrierReport>} A promise resolving to the result or rejecting with an error
     */
    ShieldInformationBarrierReportsManager.prototype.create = function (body, options, callback) {
        var queryParams = __rest(options, []), apiPath = (0, url_path_1.default)('shield_information_barrier_reports'), params = {
            qs: queryParams,
            body: body,
        };
        return this.client.wrapWithDefaultHandler(this.client.post)(apiPath, params, callback);
    };
    return ShieldInformationBarrierReportsManager;
}());
module.exports = ShieldInformationBarrierReportsManager;
